rust/src/fees.rs

```diff
@@ -0,0 +1,216 @@
+
+    let input_size = LABEL_SIZE + CBOR_TAG + ARRAY_PREFIX
+                   + tx.inputs.len() * (SMALL_ARRAY + UINT + HASH_OBJ);
+    let output_size = LABEL_SIZE + ARRAY_PREFIX
+                    + tx.outputs.len() * (SMALL_ARRAY + UINT + ADDRESS);
+    let certs_size = tx.certs.as_ref().map_or(0, |certs| certs.0.iter().map(cert_size).sum::<usize>());
+    let withdrawals_size = LABEL_SIZE + MAP_PREFIX
+                         + tx.withdrawals.as_ref().map_or(0, |wd| wd.len()) * (UINT + CREDENTIAL);
+    // TODO: updates
+    let update_size = ARRAY_PREFIX;
+    // TODO: metdata/hash
+    let metadata_size = ARRAY_PREFIX;
+    let metdata_hash_size = ARRAY_PREFIX;
+    
+    println!("input_size = {}", input_size);
+    println!("output_size = {}", output_size);
+    println!("certs_size = {}", certs_size);
+    println!("withdrawals_size = {}", withdrawals_size);
+    println!("FEE_SIZE = {}", FEE_SIZE);
+    println!("TTL_SIZE = {}", TTL_SIZE);
+    println!("update_size = {}", update_size);
+    println!("metdata_hash_size = {}", metdata_hash_size);
+    println!("witness_size = {}", witness_size);
+    println!("metadata_size = {}", metadata_size);
+    input_size + output_size + certs_size + withdrawals_size + FEE_SIZE + TTL_SIZE + update_size + metdata_hash_size + witness_size + metadata_size
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use js_chain_libs::*;
+    use address::*;
+
+    const HLEN: usize = 4;
+    const SLEN: usize = 13;
+
+    // TODO: compare raw bytes now that we have complete test vectors for this
+    // will require implementing their mock crypto though... or at least copying the outputs 
+    fn genesis_id() -> Hash {
+        Hash::new(vec![0; HLEN])
+    }
+    fn alice_key() -> PrivateKey {
+        PrivateKey::from_normal_bytes(&[228, 61, 34, 119, 224, 166, 98, 69, 109, 32, 41, 244, 193, 183, 151, 145, 1, 130, 86, 184, 181, 148, 163, 25, 206, 19, 125, 217, 15, 154, 95, 53]).unwrap()
+    }
+    fn alice_pay() -> Keyhash {
+        Keyhash::new(vec![2; HLEN])
+    }
+    fn alice_stake() -> Keyhash {
+        Keyhash::new(vec![3; HLEN])
+    }
+    fn alice_addr() -> Address {
+        BaseAddress::new(3, AddrCred::from_keyhash(alice_pay()), AddrCred::from_keyhash(alice_stake())).to_address()
+    }
+    fn bob_key() -> PrivateKey {
+        PrivateKey::from_normal_bytes(&[29, 121, 11, 180, 125, 92, 240, 44, 174, 77, 75, 175, 52, 177, 31, 232, 186, 118, 65, 184, 118, 3, 159, 236, 29, 166, 235, 108, 101, 13, 67, 36]).unwrap()
+    }
+    fn bob_pay() -> Keyhash {
+        Keyhash::new(vec![5; HLEN])
+    }
+    fn bob_stake() -> Keyhash {
+        Keyhash::new(vec![6; HLEN])
+    }
+    fn bob_addr() -> Address {
+        BaseAddress::new(3, AddrCred::from_keyhash(bob_pay()), AddrCred::from_keyhash(bob_stake())).to_address()
+    }
+
+    fn make_witnesses_vkey(tx: &TransactionBody, pks: Vec<&PrivateKey>) -> TransactionWitnessSet {
+        // TODO: implement non-mock at some point?
```

Comment:
This is just pending the tx signing test vectors, right?

Reply:
I believe so. We just were testing size at first so I didn't bother as well.